rfc/resolve-directive.md

```diff
@@ -0,0 +1,189 @@
+or we'd have to run a separate instance of the Router for internal use with a contract,
+which adds complexity again.
+
+Wouldn't it be great if we could somehow validate references in a declarative way?
+In addition, such a solution could allow us to "query" input fields for operations that are not directly related to the entity.
+
+## Solution
+
+I propose to add "internal" fields which we mark as `@inaccessible`.
+These can be used by the Router to resolve information that is useful to other Subgraphs, but not directly accessible as keys.
+
+In addition, I propose a new directive `@resolve` which allows us to define a query on an input field.
+This input field can be market as `@inaccessible` to prevent exposing it to the public API.
+As such, clients are not able to define the input field in their queries.
+Instead, the Router will execute the query defined in the `@resolve` directive and pass the result to the input.
+
+# Benefits
+
+Implicit data dependencies between services are invisible.
+If you call a mutation on Subgraph A,
+and Subgraph A needs to call Subgraph B and C to fulfill this mutation,
+the reality is that this mutation depends on Subgraph A, B, and C.
+
+However, in this scenario, we're only seeing Subgraph A in the analytics, query planning, etc. because the Subgraphs B and C are being called "off-graph".
+Implicit off-graph calls are invisible.
+We're unaware they are happening, but even worse, we simply don't understand the relationship between a root field (e.g. Mutation) and our service layer.
+
+By explicitly defining our data dependencies on the graph,
+we can build tooling around information like query plans, analytics, etc.
+so we can make these dependencies nost just visible,
+but use them to our advantage.
+
+Breaking change detection can use schema usage from these "internal" calls to prevent production issues.
+Query plans can show these implicit dependencies.
+A dashboard can show which queries depend on which services, explicit and implicit.
+Another dashboard can show dependencies between services.
+
+## Example 1: Validate a Reference
+
+```graphql
+# Review Subgraph
+type Review @key(fields: "id") {
+    id: ID!
+    rating: Int!
+    comment: String!
+}
+
+extend type Product @key(fields: "id") {
+    id: ID!
+    reviews: [Review]
+}
+
+type Query {
+    reviewByID(id: ID!): Review @inaccessible
+}
+```
+
+```graphql
+# Product Admin Subgraph
+
+input RemoveReviewInput {
+    productID: ID!
+    reviewID: ID!
+    # This field is used to validate the review exists
+    # if the Router resolves it to null, the mutation will fail because the field is not nullable
+    # if it's resolved to a value, the mutation will have access to the content as part of the input
+    review: RemoveReviewReviewInput! @inaccessible @resolve(query: "query Review($id: ID!) { reviewByID(id: $id) { id }}", variables: "{ \"id\": {{ .entity.reviewID }} }")
```

Comment:
I don't really understand this, honestly.

Why is this input necessary at all? Why isn't `@resolve` added to `@reviewID`? And whatever ID is passed by the user to `reviewID` must be found within the validation query, i.e., `reviewByID`?

Reply:
I also wonder whether this syntax is overly verbose.

Reply:
There needs to be a way to pass a header or claim, e.g. "sub", as an argument. How would you solve this?